File: src/core/candidate/candidate.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { CreateCandidateDto } from './dto/create-candidate.dto';
import { FindCandidateDto } from './dto/find-candidate.dto';
import { UpdateCandidateDto } from './dto/update-candidate.dto';
import { ICandidateRepository } from './repositories/candidate.repository';
import { Candidate } from './entities/candidate.entity';
import { Message } from '../../shared/interfaces/messages';
import { ElectionService } from '../election/election.service';

@Injectable()
export class CandidateService {
  public constructor(
    private readonly candidateRepository: ICandidateRepository,
    private readonly electionService: ElectionService,
  ) {}

  public async create(
    createCandidateDto: CreateCandidateDto,
  ): Promise<Message> {
    await this.electionService.findById(createCandidateDto.election);
    await this.candidateRepository.create(createCandidateDto);
    return { message: 'Candidato adicionado com sucesso' };
  }

  public async findById(id: string): Promise<Candidate> {
    const candidate = await this.candidateRepository.findById(id);
    if (!candidate) throw new NotFoundException('Candidato não encontrado');
    return candidate;
  }

  public findMany(
    user: string,
    findCandidateDto: FindCandidateDto,
  ): Promise<Candidate[]> {
    return this.candidateRepository.findMany(user, findCandidateDto);
  }

  public async update(
    id: string,
    updateCandidateDto: UpdateCandidateDto,
  ): Promise<Message> {
    const { election } = await this.findById(id);
    if (
      updateCandidateDto.election &&
      updateCandidateDto.election != election.id
    )
      await this.electionService.findById(updateCandidateDto.election);

    await this.candidateRepository.update(id, updateCandidateDto);
    return { message: 'Candidato atualizado com sucesso' };
  }

  public async delete(id: string): Promise<Message> {
    await this.findById(id);
    await this.candidateRepository.delete(id);
    return { message: 'Candidato removido com sucesso' };
  }
}

File: src/main/java/dev/joaopdias/vox/core/candidate/CandidateService.java
package dev.joaopdias.vox.core.candidate;

import dev.joaopdias.vox.core.candidate.dto.CandidateResponseDto;
import dev.joaopdias.vox.core.candidate.dto.CreateCandidateDto;
import dev.joaopdias.vox.core.candidate.dto.UpdateCandidateDto;
import dev.joaopdias.vox.core.candidate.entities.Candidate;
import dev.joaopdias.vox.core.election.ElectionService;
import dev.joaopdias.vox.core.election.entities.Election;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.util.UUID;
import java.util.stream.Stream;

@Service
public class CandidateService {
    private final CandidateRepository candidateRepository;

    private final ElectionService electionService;

    public CandidateService(
            CandidateRepository candidateRepository,
            ElectionService electionService
    ){
        this.candidateRepository = candidateRepository;
        this.electionService = electionService;
    }

    public CandidateResponseDto create(CreateCandidateDto createCandidateDto){
        Election election = this.electionService.findById(createCandidateDto.electionId());

        Candidate candidate = new Candidate();

        candidate.setName(createCandidateDto.name());
        candidate.setElection(election);

        this.candidateRepository.save(candidate);

        return candidate.toResponseDto();
    }

    public Candidate findById(UUID id){
        return this.candidateRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Candidato não encontrado"));
    }

    public Stream<CandidateResponseDto> findManyByElection(UUID electionId, Pageable pageable){
        this.electionService.findById(electionId);
        return this.candidateRepository
                .findManyByElectionId(electionId, pageable).stream().map(Candidate::toResponseDto);
    }

    public void update(UUID id, UpdateCandidateDto updateCandidateDto){
        Candidate candidate = this.findById(id);

        if (updateCandidateDto.name() != null) candidate.setName(updateCandidateDto.name());

        this.candidateRepository.save(candidate);
    }

    public void delete(UUID id){
        Candidate candidate = this.findById(id);
        this.candidateRepository.delete(candidate);

    }
}
